wireguard trace reason encrypted

smagnani96 opened the pull request:
This PR introduces the TRACE_REASON_ENCRYPT reason when tracing a WireGuard-encrypted packet, alongside the needed logic in the ingress path.
Each commit should contain a brief description of the changes applied.
The resulting output after applying this PR should look as follows:

```bash
# WireGuard with Overlay
1:<- endpoint 1756 flow 0x0 , identity 12670->unknown state unknown ifindex 0 orig-ip 0.0.0.0: 10.244.1.196 -> 10.244.2.22 EchoRequest
2:-> overlay flow 0x0 , identity 12670->7292 state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.244.1.196 -> 10.244.2.22 EchoRequest
3--> network encrypted  flow 0x9597ef3d , identity unknown->unknown state new ifindex eth0 orig-ip 0.0.0.0: 172.18.0.4:51871 -> 172.18.0.2:51871 udp
4-<- network encrypted  flow 0x9597ef3d , identity remote-node->unknown state new ifindex eth0 orig-ip 0.0.0.0: 172.18.0.2:51871 -> 172.18.0.4:51871 udp
5:<- overlay flow 0x0 , identity 7292->unknown state unknown ifindex cilium_vxlan orig-ip 0.0.0.0: 10.244.2.22 -> 10.244.1.196 EchoReply
6:-> endpoint 1756 flow 0x0 , identity 7292->12670 state reply ifindex lxcfe422d25f340 orig-ip 10.244.2.22: 10.244.2.22 -> 10.244.1.196 EchoReply

# WireGuard with Native Routing
1:<- endpoint 2379 flow 0x0 , identity 16125->unknown state unknown ifindex 0 orig-ip 0.0.0.0: 10.244.2.39 -> 10.244.1.72 EchoRequest
2:-> stack flow 0x0 , identity 16125->34030 state new ifindex 0 orig-ip 0.0.0.0: 10.244.2.39 -> 10.244.1.72 EchoRequest
3:-> 0: 10.244.2.39 -> 10.244.1.72 EchoRequest
4--> network encrypted  flow 0x0 , identity unknown->unknown state new ifindex eth0 orig-ip 0.0.0.0: 172.18.0.2:51871 -> 172.18.0.3:51871 udp
5-<- network encrypted  flow 0x0 , remote-node->unknown state new ifindex eth0 orig-ip 0.0.0.0: 172.18.0.3:51871 -> 172.18.0.2:51871 udp
6:<- host flow 0x0 , identity world->unknown state unknown ifindex cilium_wg0 orig-ip 0.0.0.0: 10.244.1.72 -> 10.244.2.39 EchoReply
7:-> endpoint 2379 flow 0x0 , identity 34030->16125 state reply ifindex lxc27ddb3574bfb orig-ip 10.244.1.72: 10.244.1.72 -> 10.244.2.39 EchoReply
```

Notice how lines 3-4 in Overlay and 4-5 in Native Routing have the `encrypted` flag.

This still does not solve all the Hubble/Cilium-dbg observability, as all the information concerning the inner packets (endpoint) are lost at the time of the tracing event. The purpose of this PR is to align the behavior with IPSec. See the linked issue below with the discussion.

Fixes: #34659

```release-note
Add logic to detect and trace WireGuard encrypted ingress/egress packets.
```

smagnani96 committed:
add WireGuard port to the define config map for datapath

This commit adds to the config map, containing all the defined for the
datapath, the WG_PORT value where WireGuard is listening. This value can
then be later used to match incoming network packets destinated to WireGuard.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
add logic to match WireGuard traffic

This commit introduces the logic to match WireGuard traffic, mainly for
tracing. A packet is classified as belonging to a WireGuard communication
in Cilium if all the following conditions hold true:

1. UDP packet           -- non UDP cannot be WireGuard.
2. L4 dport == WG_PORT  -- the destination is for WG_PORT, could be related
                           to a cilium-managed WireGuard communication.
3. L4 sport == dport    -- also the sender is using WG_PORT, could be
                           originated by a cilium node.
4. Identity in cluster  -- the sender is a valid node within the cluster
                           now we now for sure that is a WG-valid packet.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

jschwinger233 commented on bpf/node_config.h:
```diff
@@ -284,6 +284,7 @@ DEFINE_IPV6(HOST_IP, 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0xa, 0x
 
 #ifdef ENABLE_WIREGUARD
 # define WG_IFINDEX	42
+# define WG_PORT 51871
```
nit: use tab to align
```diff
-# define WG_PORT 51871
+# define WG_PORT	51871
```

jschwinger233 commented on bpf/lib/encrypt.h:
```diff
@@ -284,8 +288,109 @@ encrypt_overlay_and_redirect(struct __ctx_buff *ctx)
+ * - Valid node ID
+ */
+static __always_inline int
+do_decrypt(struct __ctx_buff *ctx, __u16 proto)
```
I see your point: ipsec has a `do_decrypt` so you may think wireguard should have a similar one. But I feel you just need a  `ctx_is_wireguard(__ctx_buff) bool`.

smagnani96 replied:
Agree, still would make use of the `__u16 proto` already computed.  For which file this new function would be a better fit rather than `encrypt.h`?

jschwinger233 commented on bpf/lib/encrypt.h:
```diff
@@ -284,8 +288,109 @@ encrypt_overlay_and_redirect(struct __ctx_buff *ctx)
+	}
+
+	/* Ignore invalid node ID. */
+	if (!node_id)
```
Ipsec needs node id to decrypt but not for wireguard

smagnani96 replied:
I wanted to make sure that the sender was a valid node.
Would you then just check for dport==WG_PORT? Or also make sure that sport==dport?

jschwinger233 replied:
Can we move the whole "check wireguard and notify" to https://github.com/cilium/cilium/blob/v1.17.0-pre.1/bpf/bpf_host.c#L593 in the handle_ipv4/handle_ipv6, where we have the ipcache_srcid and ip4/ip6 parsed. Over there we can call `identity_is_cluster(ipcache_srcid)` to make sure the packet is from inside the cilium.

smagnani96 replied:
Let me investigate this solution. Though, at this point `bpf_host` would have already traced `FROM_NETWORK` without the TRACE_REASON_ENCRYPTED, so we're creating a new tracing here?

jschwinger233 replied:
Good point. Therefore `do_netdev()` indeed is better choice. 
Let's just use the src port and dst port to distinguish wireguard.

smagnani96 committed:
refactoring tracing in `do_netdev` to add src identity if available

This commit refactors the tracing of the packets in the `do_netdev` function.
Before this commit, we always trace the packet with an unknown src identity.
Since few lines of code later we retrieve this value in case of an IPv4/6
packet, it makes sense to use it also while tracing the packet
so that we provide a bit more context. An example of output after applying
this patch would be as follows (no encryption, native routing):

```bash
1:<- endpoint 2223 flow 0x0 , identity 2173->unknown state unknown ifindex 0 orig-ip 0.0.0.0: 10.244.3.213 -> 10.244.1.98 EchoRequest
2:-> stack flow 0x0 , identity 2173->13606 state new ifindex 0 orig-ip 0.0.0.0: 10.244.3.213 -> 10.244.1.98 EchoRequest
3:-> 0: 10.244.3.213 -> 10.244.1.98 EchoRequest
4:<- network flow 0x0 , identity 13606->unknown state unknown ifindex eth0 orig-ip 0.0.0.0: 10.244.1.98 -> 10.244.3.213 EchoReply
5:<- host flow 0x0 , identity 13606->unknown state unknown ifindex eth0 orig-ip 0.0.0.0: 10.244.1.98 -> 10.244.3.213 EchoReply
6:-> endpoint 2223 flow 0x0 , identity 13606->2173 state reply ifindex lxcf91eb0f2f0ab orig-ip 10.244.1.98: 10.244.1.98 -> 10.244.3.213 EchoReply
```

We can now see in line 4 the source identity while logging the packet.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard tracing ingress traffic with TRACE_REASON_ENCRYPTED

This commit adds logic to trace ingress WireGuard packets with the
TRACE_REASON_ENCRYPT. The reason is to have an aligned behavior similarly
to IPSec, for which we inspect and check whether the packet is ESP and mark
it for decryption. Within WireGuard, we just need to check that it is a
valid WireGuard packet for and from a cilium-managed endpoint and trace it
accordingly without having to use ctx->mark. An example output would look
as follows:

```bash
1:<- endpoint 321 flow 0x0 , identity 41978->unknown state unknown ifindex 0 orig-ip 0.0.0.0: 10.244.2.90 -> 10.244.1.17 EchoRequest
2:-> stack flow 0x0 , identity 41978->26105 state new ifindex 0 orig-ip 0.0.0.0: 10.244.2.90 -> 10.244.1.17 EchoRequest
3:-> 0: 10.244.2.90 -> 10.244.1.17 EchoRequest
4--> network flow 0x0 , identity unknown->unknown state new ifindex eth0 orig-ip 0.0.0.0: 172.18.0.4:51871 -> 172.18.0.2:51871 udp
5-<- network encrypted  flow 0x0 , identity remote-node->unknown state new ifindex eth0 orig-ip 0.0.0.0: 172.18.0.2:51871 -> 172.18.0.4:51871 udp
6:<- host flow 0x0 , identity 26105->unknown state unknown ifindex cilium_wg0 orig-ip 0.0.0.0: 10.244.1.17 -> 10.244.2.90 EchoReply
7:-> endpoint 321 flow 0x0 , identity 26105->41978 state reply ifindex lxceef65ccc7619 orig-ip 10.244.1.17: 10.244.1.17 -> 10.244.2.90 EchoReply
```

Note how in line 5 the FROM_NETWORK packet is marked as `encrypted`.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard tracing egress traffic with TRACE_REASON_ENCRYPTED

This commit modifies the `trace.reason` in the egress path of a WireGuard
encrypted packet, to trace it with TRACE_REASON_ENCRYPT.
An example output would look as follows:

```bash
1:<- endpoint 321 flow 0x0 , identity 41978->unknown state unknown ifindex 0 orig-ip 0.0.0.0: 10.244.2.90 -> 10.244.1.17 EchoRequest
2:-> stack flow 0x0 , identity 41978->26105 state new ifindex 0 orig-ip 0.0.0.0: 10.244.2.90 -> 10.244.1.17 EchoRequest
3:-> 0: 10.244.2.90 -> 10.244.1.17 EchoRequest
4--> network encrypted flow 0x0 , identity unknown->unknown state new ifindex eth0 orig-ip 0.0.0.0: 172.18.0.4:51871 -> 172.18.0.2:51871 udp
5-<- network encrypted  flow 0x0 , identity remote-node->unknown state new ifindex eth0 orig-ip 0.0.0.0: 172.18.0.2:51871 -> 172.18.0.4:51871 udp
6:<- host flow 0x0 , identity 26105->unknown state unknown ifindex cilium_wg0 orig-ip 0.0.0.0: 10.244.1.17 -> 10.244.2.90 EchoReply
7:-> endpoint 321 flow 0x0 , identity 26105->41978 state reply ifindex lxceef65ccc7619 orig-ip 10.244.1.17: 10.244.1.17 -> 10.244.2.90 EchoReply
```

Note how also in line 4 the TO_NETWORK packet is marked as `encrypted`.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
refactoring unneeded if condition inside ifdef in `do_netdev`

This commit moves the check `from_host` inside the ifdef that would use it,
allowing us to save unneeded complexity when using
`ENABLE_HOST_FIREWALL && !ENABLE_MASQUERADE_IPV4`.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

dylandreimerink reviewed (approved):
LGTM

jschwinger233 reviewed (approved):
Awesome! Thank you!

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1156,15 +1161,28 @@ do_netdev(struct __ctx_buff *ctx, __u16 proto, const bool from_host)
+#ifdef ENABLE_WIREGUARD
+		if (!from_host) {
+			next_proto = ip6->nexthdr;
+			l4_off = ETH_HLEN + ipv6_hdrlen(ctx, &next_proto);
```
Note that `ipv6_hdrlen()` can return errors. Which we should probably tolerate in this case (since it can be arbitrary inbound traffic), but not continue in the WG check (since we wouldn't have extracted the actual L4 proto.).

smagnani96 replied:
Correct, let me know in case we want to address it here, thanks for it!

smagnani96 replied:
Resolving conversation, proposing fix in https://github.com/cilium/cilium/pull/35345.

julianwiedmann reviewed (approved):
Nice work @smagnani96!

There's a tiny bit we'll need to fix-up, but I'll resolve the comment so we can merge this and fix with a follow-up PR.

kaworu commented on bpf/bpf_host.c:
```diff
@@ -1156,15 +1161,28 @@ do_netdev(struct __ctx_buff *ctx, __u16 proto, const bool from_host)
+		if (!from_host) {
+			next_proto = ip6->nexthdr;
+			l4_off = ETH_HLEN + ipv6_hdrlen(ctx, &next_proto);
+			if (ctx_is_wireguard(ctx, l4_off, next_proto, ipcache_srcid))
+				trace.reason = TRACE_REASON_ENCRYPTED;
```
@learnitall with https://github.com/cilium/cilium/pull/29616 in mind for context, do we need to add `trace.monitor=0` here? If yes, then shouldn't this line be caught by `encrypt_agg.cocci`? As far as I understand it's not triggering the cocci patch because there is no explicit `trace.monitor=${something-non-zero}` following this line.

smagnani96 replied:
I had the [same concern](https://github.com/cilium/cilium/issues/34659#:~:text=The%20encrypt_agg.cocci%20has%20been%20introduced%20to%20make%20sure%20that%20no%20aggregation%20is%20performed%20when%20tracing%20encrypted%20packet%20(IPSec).%20However%2C%20when%20applying%20the%20patch%20hence%20tracing%20Ingress%20Wireguard%20as%20encrypted%2C%20the%20cil_from_netdev%20program%20uses%20the%20TRACE_PAYLOAD_LEN%20macro.). Differently than IPSec, these traces would be generated anyway independently by us marking with TRACE_REASON_ENCRYPTED or not. For this reason, here I didn't want to loose the TRACE_PAYLOAD_LEN value.
I also agree that what you describe is the reason why  `encrypt_agg.cocci` is not detecting this. Let me know in case anything else is needed.

learnitall replied:
> shouldn't this line be caught by `encrypt_agg.cocci`?

It _should_ but this usage pattern of `trace_ctx` wasn't added. We can look into adding another case which requires that `trace.monitor = 0` follows `trace.reason = TRACE_REASON_ENCRYPTED`.
 
> Differently than IPSec, these traces would be generated anyway independently by us marking with TRACE_REASON_ENCRYPTED or not.

Can you expand on what you mean? I understand that the traces would be generated anyway regardless of if the packet was WG or not, but I don't understand what you mean by 'differently than IPSec'.

Overall, the reason why we've paired setting `monitor=0` with `TRACE_REASON_ENCRYPTED` is we've assumed that anytime a `TRACE_REASON_ENCRYPTED` event is sent, the packet the event was generated for lacks the connection tracking information to perform event aggregation, and as a result, these events produce a lot of noise. My understand of the specifics of the aggregation techniques and how we handle WG traffic isn't developed enough to provide a 'yes' or 'no' here, so I'd recommend we experiment:

1. Can we compare the observed events for WG traffic before and after the PR? If this PR dramatically increases the number of events, then we should set `monitor=0`. For example, here's a before/after comparison of the effect of https://github.com/cilium/cilium/pull/29616 for IPSec using `hubble observe --follow`:

Before:

```
Feb 22 22:21:59.069: default/test2:53538 (ID:41889) -> kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-stack FORWARDED (UDP)
Feb 22 22:21:59.069: default/test2:53538 (ID:41889) <> kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-stack FORWARDED (UDP)
Feb 22 22:21:59.069: default/test2:53538 (ID:41889) <> kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-stack FORWARDED (UDP)
Feb 22 22:21:59.069: default/test2:53538 (ID:41889) -> kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-endpoint FORWARDED (UDP)
Feb 22 22:21:59.070: default/test2:53538 (ID:41889) <- kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-stack FORWARDED (UDP)
Feb 22 22:21:59.071: kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) <> default/test2:53538 (ID:41889) to-stack FORWARDED (UDP)
Feb 22 22:21:59.071: kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) <> default/test2:53538 (ID:41889) to-stack FORWARDED (UDP)
Feb 22 22:21:59.071: default/test2:53538 (ID:41889) <- kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-endpoint FORWARDED (UDP)
Feb 22 22:21:59.072: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: SYN)
Feb 22 22:21:59.072: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: SYN)
Feb 22 22:21:59.072: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: SYN)
Feb 22 22:21:59.072: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: SYN, ACK)
Feb 22 22:21:59.073: default/nginx:80 (ID:33458) <> default/test2:49594 (ID:41889) to-stack FORWARDED (TCP Flags: SYN, ACK)
Feb 22 22:21:59.073: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: SYN, ACK)
Feb 22 22:21:59.073: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.073: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.073: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/nginx:80 (ID:33458) <> default/test2:49594 (ID:41889) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/nginx:80 (ID:33458) <> default/test2:49594 (ID:41889) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/nginx:80 (ID:33458) <> default/test2:49594 (ID:41889) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.076: default/nginx:80 (ID:33458) <> default/test2:49594 (ID:41889) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
```

After:

```
Feb 22 22:26:47.377: default/test:47138 (ID:40856) -> kube-system/kube-dns-6f9b8847ff-hqbpb:53 (ID:28365) to-stack FORWARDED (UDP)
Feb 22 22:26:47.377: default/test:47138 (ID:40856) -> kube-system/kube-dns-6f9b8847ff-hqbpb:53 (ID:28365) to-endpoint FORWARDED (UDP)
Feb 22 22:26:47.378: default/test:47138 (ID:40856) <- kube-system/kube-dns-6f9b8847ff-hqbpb:53 (ID:28365) to-stack FORWARDED (UDP)
Feb 22 22:26:47.378: default/test:47138 (ID:40856) <- kube-system/kube-dns-6f9b8847ff-hqbpb:53 (ID:28365) to-endpoint FORWARDED (UDP)
Feb 22 22:26:47.379: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: SYN)
Feb 22 22:26:47.379: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: SYN)
Feb 22 22:26:47.379: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: SYN, ACK)
Feb 22 22:26:47.380: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK)
Feb 22 22:26:47.380: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: SYN, ACK)
Feb 22 22:26:47.380: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:26:47.380: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:26:47.380: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:26:47.381: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:26:47.381: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:26:47.383: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:26:47.383: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:26:47.392: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:26:47.392: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:26:47.392: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:26:47.392: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK)
```

2. Can we try setting `monitor=0` here and observe the impact on the generated events? If there's information missing or the impact isn't noticeable, then we can pass on it.

@smagnani96 would you have time to try these tests out? I can help out as well if you'd like.

kaworu replied:
@learnitall I talked with Simone and he helped me understand that changing the trace reason for wg in this PR shouldn't affect monitor aggregation.

The two changes in `do_netdev` (this thread and [L1223](https://github.com/cilium/cilium/pull/35183/files/785a3c9f94dab5457371db8c308b839c40153f2f#r1795537643)) are in the ingress path, and the observation point will consistently be `TRACE_FROM_NETWORK` in the wg case. `TRACE_FROM_*` events will not be emitted unless monitor aggregation is set to "none" _regardless of the `monitor` value given to `send_trace_notify`_, see [here](https://github.com/cilium/cilium/blob/253a05bef38306624eb6675c16a4e90cb3e5c729/bpf/lib/trace.h#L165-L177).

[`cil_to_netdev`](https://github.com/cilium/cilium/pull/35183/files/785a3c9f94dab5457371db8c308b839c40153f2f#r1795543959) is in the egress path, but there we go through[ `handle_to_netdev_ipv{6,4}`](https://github.com/cilium/cilium/blob/253a05bef38306624eb6675c16a4e90cb3e5c729/bpf/bpf_host.c#L1443-L1454) and do the ct lookup.

Thus, I marked my comments as resolved, but feel free to respond if you think we've missed something :pray:

learnitall replied:
Ok sweet! That makes a ton of sense to me. Thank you!

smagnani96 replied:
Thanks to you both :smiley: 
As mentioned by @kaworu, this patch shouldn't lead to logging more events than we would normally expect:

1. ingress: `TRACE_FROM_NETWORK` is not logged by default when `monitorAggregation > none`
2. egress: `trace.monitor` is modified only by a successful `ct_lookup`, which shouldn't happen when dealing with the encrypted packet, thus not logging the event when `monitorAggregation > none`.

For instance, with the default `monitorAggregation=medium` I see only the following packets:

```bash
#request
-> stack flow 0x0 , identity 19646->4691 state new ifindex 0 orig-ip 0.0.0.0: 10.244.1.178 -> 10.244.2.58 EchoRequest
-> 0: 10.244.1.178 -> 10.244.2.58 EchoRequest
#response
-> endpoint 388 flow 0x0 , identity 4691->19646 state reply ifindex lxc2e2524a80f9c orig-ip 10.244.2.58: 10.244.2.58 -> 10.244.1.178 EchoReply
```

julianwiedmann replied:
>egress: trace.monitor is modified only by a successful ct_lookup, which shouldn't happen when dealing with the encrypted packet, thus not logging the event when monitorAggregation > none.

This is related to this code section, right?
https://github.com/cilium/cilium/blob/a2cf84ec76daa6c12950f7f0efa7948d7d904f30/bpf/bpf_host.c#L1527

There's two aspects to consider
1. when the HostFW is enabled, the preceding Egress-policy check would have already performed a CT lookup (since encrypted WG traffic looks like host-originating traffic), and touched the `trace` struct. Arguably we should be smarter here, and as https://github.com/cilium/cilium/issues/32335 suggests not apply HostFW to our own WG traffic (or are we already doing that? Should be straight-forward to repro).
2. when BPF masq is enabled, we apply CT for [host-originating traffic like WG](https://github.com/cilium/cilium/issues/34089) (to avoid port collisions with SNATed pod-originating connections). Here `handle_nat_fwd()` would touch the `trace` struct.

smagnani96 replied:
Hey @julianwiedmann, thanks for the feedback!
Your comment made me realize that right now we might lose the encryption bit in that codepath, such as in case of `handle_nat_fwd()`. I'm trying to tackle those cases in #35354, any further comment is appreciated :smiley:

kaworu commented on bpf/bpf_host.c:
```diff
@@ -1186,15 +1204,28 @@ do_netdev(struct __ctx_buff *ctx, __u16 proto, const bool from_host)
+			next_proto = ip4->protocol;
+			l4_off = ETH_HLEN + ipv4_hdrlen(ip4);
+			if (ctx_is_wireguard(ctx, l4_off, next_proto, ipcache_srcid))
+				trace.reason = TRACE_REASON_ENCRYPTED;
```
ditto `trace.monitor=0`

kaworu commented on bpf/bpf_host.c:
```diff
@@ -1489,12 +1522,14 @@ int cil_to_netdev(struct __ctx_buff *ctx __maybe_unused)
 	 * encrypted WireGuard UDP packets), we check whether the mark
 	 * is set before the redirect.
 	 */
+	trace.reason = TRACE_REASON_ENCRYPTED;
```
ditto `trace.monitor=0` but this one is less clear to me.
